File: backend/ts_template/devdonalds.ts
import express from "express";
import type { Request, Response } from "express";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  type: "recipe";
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  type: "ingredient";
  cookTime: number;
}

interface RecursiveState {
  name: string;
  cookTime: number;
  ingredients: Map<string, number>
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

let cookbook: Map<string, Map<any, any>> = new Map();

app.post("/clear", (req: Request, res: Response) => {
  cookbook = new Map();
  return res.json({});
});

// Task 1 helper (don't touch)
app.post("/parse", (req: Request, res: Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  }
  res.json({ msg: parsed_string });
  return;

});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  let buff = [];
  recipeName = recipeName.replace(/[^a-z_\- ]/gi, "");
  for (const c of recipeName) {
    if ("_- ".includes(c)) {
      if (!buff.length || buff[buff.length - 1] == " ") {
        continue
      }
      buff.push(" ");
    } else if (!buff.length || buff[buff.length - 1] == " ") {
      buff.push(c.toUpperCase());
    } else if (buff[buff.length - 1] != " ") {
      buff.push(c.toLowerCase());
    }
  }
  if (!buff.length) {
    return null;
  }
  return buff.join("").trim();
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req: Request, res: Response) => {
  const status = handle_create_recipe(req.body);
  if (!status) {
    return res.status(400).json({ error: "Failed to save entry." })
  }
  return res.json({})
});

const handle_create_recipe = (entry: recipe | ingredient): boolean => {
  if (cookbook[entry.name]) {
    return false;
  }
  if (entry.type == "recipe") {
    let set_ = new Set(
      (entry as recipe).requiredItems.map(v => v.name)
    );
    if (set_.size !== entry.requiredItems.length) {
      return false;
    }
  } else if (entry.type == "ingredient") {
    if (entry.cookTime < 0) {
      return false;
    }
  } else {
    return false;
  }
  cookbook[entry.name] = entry;
  return true;
}

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req: Request, res: Response) => {
  const name = req.query.name as string;
  const state: RecursiveState = handle_get_recipe(name, { name, cookTime: 0, ingredients: new Map() });

  if (!state) {
    return res.status(400).json({ error: "Failed to load recipe." })
  }
  return res.json({
    name: state.name,
    cookTime: state.cookTime,
    // .map's parameter is array [key, value]
    ingredients: Object.entries(state.ingredients).map((t) => {
      return { name: t[0], quantity: t[1] }
    })
  })
});

const handle_get_recipe = (name: string, state: RecursiveState, quantity: number = 1) => {
  const entry: ingredient | recipe = cookbook[name];
  if (!entry) {
    return null;
  }
  if (entry.type === "ingredient") {
    if (entry.name == state.name) {
      return null;
    }
    state.cookTime += entry.cookTime * quantity;
    state.ingredients[name] = (state.ingredients[name] ?? 0) + quantity;
    return state;
  }
  for (const ingr of entry.requiredItems) {
    let recipe = handle_get_recipe(ingr.name, state, quantity * ingr.quantity);
    if (!recipe) {
      return null;
    }
  }
  return state;
}

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});

File: backend/py_template/devdonalds.py
import string

from dataclasses import dataclass
from typing import List, Literal, TypedDict
from flask import Flask, request, jsonify

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str

@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int

@dataclass
class RecipeEntry(CookbookEntry):
	type: Literal["recipe"]
	name: str
	requiredItems: list[RequiredItem]

@dataclass
class IngredientEntry(CookbookEntry):
	type: Literal["ingredient"]
	name: str
	cookTime: int

class ReturnIngredient(TypedDict):
	name: str
	quantity: int

class RecursiveState(TypedDict):
	name: str
	cookTime: int
	ingredients: dict[str, int]

# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook: dict[str, IngredientEntry | RecipeEntry] = {}

# Endpoint to clear the cookbook
@app.route("/clear", methods=['POST'])
def clear():
	global cookbook
	cookbook = {}
	return jsonify({}), 200

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that 
def parse_handwriting(recipeName: str) -> str | None:
	buff = []
	valid_chars = string.ascii_letters + " "
	for c in recipeName:
		if c in "_- ":
			if not buff or buff[-1] == " ":
				continue
			buff.append(" ")
		elif c not in valid_chars:
			continue
		elif not buff or buff[-1] == " ":
			buff.append(c.upper())

		elif buff[-1] != " ":
			buff.append(c.lower())

	if not len(buff):
		return None
	return "".join(buff).strip()


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():
	data = request.get_json()
	status = handle_create_recipe(data)

	if not status:
		return {"error": "Failed to save entry."}, 400
	return jsonify({}), 200


def handle_create_recipe(entry: dict) -> bool:
	if cookbook.get(entry["name"]):
		return False
	model = None
	if entry["type"] == "recipe":
		entry["requiredItems"] = [RequiredItem(**i) for i in entry.get("requiredItems", [])]
		model = RecipeEntry(**entry)
		# We can also do set_ = set([item.name for item in model.requiredItems])
		# and compare the length, but this is also a solution.
		set_ = set()
		# O(n) time complexity since item lookup in set is O(1)
		for item in model.requiredItems:
			if item.name in set_:
				return False 
			set_.update({item.name})

	elif entry["type"] == "ingredient":
		model = IngredientEntry(**entry)
		if model.cookTime < 0:
			return False
	else:
		return False

	cookbook[entry["name"]] = model
	return True

# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
	name = request.args["name"]
	state = handle_get_recipe(name, {"name": name, "cookTime": 0, "ingredients": {}})

	if not state:
		return {"error": "Failed to load recipe."}, 400

	return jsonify({
		"name": state["name"],
		"cookTime": state["cookTime"],
		"ingredients": [{"name": k, "quantity": v} for k, v in state["ingredients"].items()]
	}), 200

def handle_get_recipe(name: str, state: RecursiveState, quantity: int = 1) -> RecursiveState | None:
	entry = cookbook.get(name)
	if not entry:
		return None
	if entry.type == "ingredient":
		# This is the searched recipe
		if entry.name == state["name"]:
			return None
		state["cookTime"] += entry.cookTime * quantity
		state["ingredients"][name] = state["ingredients"].get(name, 0) + quantity
		return state
	for ingr in entry.requiredItems:
		recipe = handle_get_recipe(name=ingr.name, state=state, quantity=quantity * ingr.quantity)
		# Fail everything if any recipe is not in the cookbook.
		if recipe is None:
			return None
	return state



# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)
